File: src/main/scala/no/bekk/scala/exercises/Ex5_CompetingTypes.scala
package no.bekk.scala.exercises

import no.bekk.scala.model.Car

object Ex5_CompetingTypes {

	trait Competing[T] {
		def compete(a: T, b: T): T
	}

	object Competing {
		implicit object CompetingCar extends Competing[Car] {
			def compete(a: Car, b: Car) =
				if (a.hp > b.hp) a else b
		}
	}

	def compete[A](competitors: List[A])(implicit ev: Competing[A]) =
		competitors.reduce(ev.compete)
}

File: src/main/java/no/bekk/java/exercises/Ex5_CompetingTypes.java
package no.bekk.java.exercises;

import no.bekk.java.model.Car;

import java.util.List;

import static no.bekk.java.util.Utils.implementThis;

interface Competing<T> {
	T get();
	Competing<T> competeWith(Competing<T> other);
}
public class Ex5_CompetingTypes {
	static public <T> T compete(List<Competing<T>> competitors) {
		return implementThis();
	}
}

class CarAdapter implements Competing<Car> {
	private Car car;

	CarAdapter(Car car) {
		this.car = car;
	}

	public Car get() {
		return implementThis();
	}

	public Competing<Car> competeWith(Competing<Car> other) {
        return implementThis();
	}
}